Sources/Datadog/URLSessionAutoInstrumentation/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,205 @@
+        }
+        self.dataTaskWithURLRequestAndCompletion = try DataTaskWithURLRequestAndCompletion(interceptor: interceptor)
+        self.dataTaskWithURLRequest = try DataTaskWithURLRequest(interceptor: interceptor)
+        self.interceptor = interceptor
+    }
+
+    func swizzle() {
+        dataTaskWithURLRequestAndCompletion.swizzle()
+        dataTaskWithURLAndCompletion?.swizzle()
+        dataTaskWithURLRequest.swizzle()
+        dataTaskWithURL?.swizzle()
+    }
+
+    // MARK: - Swizzlings
+
+    typealias CompletionHandler = (Data?, URLResponse?, Error?) -> Void
+
+    /// Swizzles the `URLSession.dataTask(with:completionHandler:)` for `URLRequest`.
+    class DataTaskWithURLRequestAndCompletion: MethodSwizzler<
+        @convention(c) (URLSession, Selector, URLRequest, CompletionHandler?) -> URLSessionDataTask,
+        @convention(block) (URLSession, URLRequest, @escaping CompletionHandler) -> URLSessionDataTask
+    > {
+        private static let selector = #selector(
+            URLSession.dataTask(with:completionHandler:) as (URLSession) -> (URLRequest, @escaping CompletionHandler) -> URLSessionDataTask
+        )
+
+        private let method: FoundMethod
+        private let interceptor: URLSessionInterceptorType
+
+        init(interceptor: URLSessionInterceptorType) throws {
+            self.method = try Self.findMethod(with: Self.selector, in: URLSession.self)
+            self.interceptor = interceptor
+            super.init()
+        }
+
+        func swizzle() {
+            typealias Signature = @convention(block) (URLSession, URLRequest, CompletionHandler?) -> URLSessionDataTask
+            swizzle(method) { previousImplementation -> Signature in
+                return { [weak self] session, urlRequest, completionHandler -> URLSessionDataTask in
+                    let task: URLSessionDataTask
+
+                    if completionHandler != nil {
+                        var taskReference: URLSessionDataTask?
+
+                        let newCompletionHandler: CompletionHandler = { data, response, error in
+                            if let task = taskReference { // sanity check, should always succeed
+                                self?.interceptor.taskCompleted(urlSession: session, task: task, error: error)
+                            }
+                            completionHandler?(data, response, error)
+                        }
+
+                        let newRequest = self?.interceptor.modify(request: urlRequest) ?? urlRequest
+
+                        task = previousImplementation(session, Self.selector, newRequest, newCompletionHandler)
+                        taskReference = task
+                    } else {
+                        // The `completionHandler` is `nil` on iOS 11 or 12, because the `dataTask(with:)` (for `URL` and `URLRequest`)
+                        // calls the `dataTask(with:completionHandler:)` (for `URLRequest`) internally by nullifying the completion block.
```

Comment:
Not sure if I understand 🤔. This `completionHandler == nil` means "iOS 11 or 12 calling this method from `dataTask(with:)`". In such case, we don't want to notify the task completion (`interceptor.taskCompleted(...)`), because it will be delivered on the `URLSessionDelegate`.